WalletWasabi/Services/WalletService.cs

```diff
@@ -538,198 +563,7 @@ private async Task<bool> ProcessBlockAsync(Height height, Block block, IEnumerab
-					}
-				}
-			}
-
-			List<SmartCoin> spentOwnCoins = null;
-
-			for (var i = 0U; i < tx.Transaction.Outputs.Count; i++)
-			{
-				// If transaction received to any of the wallet keys:
-				var output = tx.Transaction.Outputs[i];
-				HdPubKey foundKey = KeyManager.GetKeyForScriptPubKey(output.ScriptPubKey);
-				if (foundKey != default)
-				{
-					walletRelevant = true;
-
-					foundKey.SetKeyState(KeyState.Used, KeyManager);
-					if (output.Value <= ServiceConfiguration.DustThreshold)
-					{
-						continue;
-					}
-
-					spentOwnCoins = spentOwnCoins ?? Coins.Where(x => tx.Transaction.Inputs.Any(y => y.PrevOut.Hash == x.TransactionId && y.PrevOut.N == x.Index)).ToList();
-					var anonset = tx.Transaction.GetAnonymitySet(i);
-					if (spentOwnCoins.Count != 0)
-					{
-						anonset += spentOwnCoins.Min(x => x.AnonymitySet) - 1; // Minus 1, because do not count own.
-					}
-
-					SmartCoin newCoin = new SmartCoin(txId, i, output.ScriptPubKey, output.Value, tx.Transaction.Inputs.ToTxoRefs().ToArray(), tx.Height, tx.IsRBF, anonset, isLikelyCoinJoinOutput, foundKey.Label, spenderTransactionId: null, false, pubKey: foundKey); // Do not inherit locked status from key, that's different.
-
-					if (Coins.TryAdd(newCoin))
-					{
-						TransactionCache.TryAdd(tx);
-
-						// If it's being mixed and anonset is not sufficient, then queue it.
-						if (newCoin.Unspent && ChaumianClient.HasIngredients && newCoin.AnonymitySet < ServiceConfiguration.MixUntilAnonymitySet && newCoin.IsLikelyCoinJoinOutput && ChaumianClient.State.Contains(newCoin.SpentOutputs))
-						{
-							try
-							{
-								await ChaumianClient.QueueCoinsToMixAsync(newCoin);
-							}
-							catch (Exception ex)
-							{
-								Logger.LogError<WalletService>(ex);
-							}
-						}
-
-						// Make sure there's always 21 clean keys generated and indexed.
-						KeyManager.AssertCleanKeysIndexed(isInternal: foundKey.IsInternal);
-
-						if (foundKey.IsInternal)
-						{
-							// Make sure there's always 14 internal locked keys generated and indexed.
-							KeyManager.AssertLockedInternalKeysIndexed(14);
-						}
-					}
-					else // If we had this coin already.
-					{
-						if (newCoin.Height != Height.Mempool) // Update the height of this old coin we already had.
-						{
-							SmartCoin oldCoin = Coins.FirstOrDefault(x => x.TransactionId == txId && x.Index == i);
-							if (oldCoin != null) // Just to be sure, it is a concurrent collection.
-							{
-								oldCoin.Height = newCoin.Height;
-							}
-						}
-					}
-				}
-			}
-
-			// If spends any of our coin
-			for (var i = 0; i < tx.Transaction.Inputs.Count; i++)
-			{
-				var input = tx.Transaction.Inputs[i];
-
-				var foundCoin = Coins.FirstOrDefault(x => x.TransactionId == input.PrevOut.Hash && x.Index == input.PrevOut.N);
-				if (foundCoin != null)
-				{
-					walletRelevant = true;
-					ChaumianClient.ExposedLinks.TryRemove(foundCoin.GetTxoRef(), out _);
```

Comment:
In the refactored code this dependency was removed. The `WalletService` is subscribed to the `CoinSpentOrSpenderConfirmed` event (this event is not new) and then it removes the exposed link.